architecture_records/002-acceleration-framework.md

```diff
@@ -0,0 +1,302 @@
+
+    def loading(model_path: str, **kwargs):
+        pass
+
+    # augment model or accelerator object
+    def augmentation(model: nn.Module, **kwargs):
+        # accelerator = kwargs.get('accelerator') # if needed for configs (e.g. FSDP)
+        pass
+
+    def callbacks(model: nn.Module, **kwargs):
+        cbks = []
+        return cbks
+```
+
+Even though they are all optional, at least one out of the three should be implemented.
+
+### Dependency Management
+
+Take note:
+- all plugin deps must be enforced to be optional deps in `pyproject.toml`, see [116](https://github.com/foundation-model-stack/fms-hf-tuning/pull/116). If the dep is not installed, and the plugin is enabled, raise exception.
+- any plugin that requires CUDA build tools (e.g. `triton` kernels) will need to be run in with [CUDA Toolkit dependencies (see this link for an example of a Debian installation)](https://developer.nvidia.com/cuda-12-2-0-download-archive?target_os=Linux&target_arch=x86_64&Distribution=Debian&target_version=11&target_type=deb_local). 
+    * whenever CUDA is needed, the framework will check for the CUDA_TOOLS dependency.
+
+### Minimal and Controlled Changes to Training Script
+
+All proposed code changes to `sft_trainer.py` contained in minimal lines of code:
+- Plugins loaded by discovery; transparent to `sft_trainer.py`.
+- Plugin configuration automatically parsed.
+- Passthrough to original operation if `Framework` is disabled.
+
+```python
+from tuning.acceleration import AccelerationFramework
+
+# Minor Change 1: creating the framework object
+framework = None
+if framework_args.config_file is not None:
+    framework = AccelerationFramework(framework_args.config_file)
+
+# Minor Change 2: custom loader (if necessary)
+_model_loader = AutoModelForCausalLM.from_pretrained # default
+if framework is not None and framework.requires_custom_loading:
+    _model_loader = framework.model_loader # drop in replacement
+
+# will passthrough the default loader if framework is disabled
+model = _model_loader(
+    model_args.model_name_or_path,
+    cache_dir=train_args.cache_dir,
+    torch_dtype=get_torch_dtype(model_args.torch_dtype),
+    attn_implementation="flash_attention_2" if model_args.use_flash_attn else None,
+)
+
+# Minor Change 3: 
+if framework is not None and framework.requires_agumentation:
+    # will also take in some other configs that may affect augmentation
+    # some of these args may be modified due to the augmentation
+    # e.g., peft_config will be consumed in augmentation, and returned as None 
+    #       to prevent SFTTrainer from doing extraneous PEFT logic
+    model, (peft_config,) = framework.augmentation(
+        model, trainer.accelerator,
+        train_args, modifiable_args=(peft_config,),
+    )
+
+# instantiate trainer. Pass in model (with training enchancements)
+trainer = Trainer(model, ...)
+
+# Minor Change 4: add trainer callbacsk
+trainer.add_callbacks(framework.callbacks())
```

Comment:
isn't `Trainer` part of HF transformers? does it have an `add_callbacks` method and, if not, how do we add that?

Reply:
Yes! the `Trainer` aluded to here indeed is the `tranformers.Trainer`, which does [indeed have an `add_callback` method](https://github.com/huggingface/transformers/blob/e34da3ee3c9d2d628fdbeb60cee45c4f8f32945a/src/transformers/trainer.py#L718) (sorry there was a minor typo).

And as you know `trl.SFTTrainer` inherits from `tranformers.Trainer`,